BEIMA.Backend/ReportService/ReportWriter.cs

```diff
@@ -0,0 +1,286 @@
+        /// file containes a header and then the data of all devices
+        /// related to the specific deviceTypeId. If the device type id
+        /// doesn't exist, return null, 
+        /// </summary>
+        /// <returns></returns>
+        public byte[] GeneratDeviceReportByDeviceType(ObjectId deviceTypeId)
+        {
+            var db = MongoDefinition.MongoInstance;
+
+            // Check if device type exists
+            var deviceTypeDocument = db.GetDeviceType(deviceTypeId);
+            if(deviceTypeDocument == null)
+            {
+                return null;
+            }
+
+            // Get all devices associated with the device type
+            var deviceType = BsonSerializer.Deserialize<DeviceType>(deviceTypeDocument);
+            var deviceFilter = MongoFilterGenerator.GetEqualsFilter("deviceTypeId", deviceType.Id);
+            var deviceDocuments = db.GetFilteredDevices(deviceFilter);
+
+            byte[] fileBytes;
+            using(var stream = new MemoryStream())
+            using (var csvWriter = new StreamWriter(stream))
+            {
+                var headers = GenerateDeviceHeaders(deviceType);
+                if (deviceDocuments.Count == 0)
+                {
+                    csvWriter.Write(headers);
+                }
+                else
+                {
+                    csvWriter.WriteLine(headers);
+                }
+
+                // Write all device data                            
+                for (var i = 0; i < deviceDocuments.Count; i++)
+                {
+                    // Write device's values
+                    var deviceDoc = deviceDocuments[i];
+                    var device = BsonSerializer.Deserialize<Device>(deviceDoc);
+                    var values = GenerateDeviceValues(device, deviceType);
+
+                    // Ensure newline isn't added on last record
+                    if (i != deviceDocuments.Count - 1)
+                    {
+                        csvWriter.WriteLine(values);
+                    }
+                    else
+                    {
+                        csvWriter.Write(values);
+                    }
+                }
+                // Flush writer and copy data to byte[]
+                csvWriter.Flush();
+                fileBytes = stream.ToArray();
+            }
+            return fileBytes;
+        }
+
+        /// <summary>
+        /// Generates a byte[] containing a zip file with a file entry
+        /// per device type. Each of the file entries will contains
+        /// a header and then the data of all devices related to that specific
+        /// device type. If there are no device types returns then null is returned.
+        /// A file entry may contain only the header row if no device is associated
+        /// with that specific device type
+        /// </summary>
+        /// <returns></returns>
+        public byte[] GenerateAllDeviceReports()
```

Comment:
I think that could be a good idea, having the ability to generate a table based off only the mandatory fields so that we can put all of the devices onto it.

Reply:
I will add this to a separate task in order to keep the pr size down